### PR TITLE
ENH: cupyx/scipy/signal: add missing LTI format conversions

### PR DESCRIPTION
**Pull request by ev-br**
cross-ref gh-7403

- [x] tf2zpk
- [x] tf2sos
- [x] tf2ss
- [x] ss2tf
- [x] zpk2ss
- [x] ss2zpk
- [x] sos2tf
- [x] sos2zpk
- [x] BadCoefficients

Note that there are two utilities which delegate to NumPy on CPU, `_polyutils.roots` and `.poly`. Both rely on `np.linalg.eigvals`, which on GPU is only available from MAGMA.

<s>this is on top of gh-7644</s>

### CODE REVIEW

**Comment by ev-br** [comment]
Rebased on main, this PR is ready for review.

File: cupyx/scipy/signal/_polyutils.py [hunk 1]
```diff
@@ -15,6 +15,31 @@ def roots(arr):
...
+    dt = seq_of_zeros.dtype
+    a = cupy.ones((1,), dtype=dt)
+    for zero in seq_of_zeros:
+        a = cupy.convolve(a, cupy.r_[1, -zero], mode='full')
+
+    if issubclass(a.dtype.type, cupy.complexfloating):
+        # if complex roots are all complex conjugates, the roots are real.
+        roots = cupy.asarray(seq_of_zeros, dtype=complex)
+        if cupy.all(cupy.sort(roots) == cupy.sort(roots.conjugate())):
+            a = a.real.copy()
```
Comment (takagi) [inline cupyx/scipy/signal/_polyutils.py:38]: How about computing on CPU not only `eigvals` but also the whole of these calculations, then `cupy.asarray` just before return?

File: tests/cupyx_tests/scipy_tests/signal_tests/test_iir_filter_conversion.py [hunk 2]
```diff
@@ -355,6 +355,195 @@ def test_bad_args(self):
...
+
+    @testing.numpy_cupy_allclose(scipy_name='scp')
+    def test_basic(self, xp, scp):
+        sos = xp.array([[1, 1, 1, 1, 0, -1],
+                        [-2, 3, 1, 1, 10, 1]])
+        b, a = scp.signal.sos2tf(sos)
+        return b, a
+
+        assert_array_almost_equal(b, [-2, 1, 2, 4, 1])
+        assert_array_almost_equal(a, [1, 10, 0, -10, -1])
```
Comment (takagi) [inline tests/cupyx_tests/scipy_tests/signal_tests/test_iir_filter_conversion.py:544]: Would you remove these unreachable lines?

**Review by takagi — COMMENTED** [review]
LGTM except for some comments!

**Comment by ev-br** [comment]
Comments addressed.

**Review by takagi — APPROVED** [review]

**Comment by takagi** [comment]
Thanks, @ev-br!

**Comment by ev-br** [comment]
Thank you for the reviews! (this PR and others)

**Pull request merged**